models/classes/runner/navigation/QtiRunnerNavigation.php

```diff
@@ -101,41 +108,72 @@ public static function move($direction, $scope, RunnerServiceContext $context, $
+    public static function checkTimedSectionExit(RunnerServiceContext $context, $nextPosition): void
     {
         $timerConfig = $context->getTestConfig()->getConfigValue('timer');
+
         if (empty($timerConfig['keepUpToTimeout'])) {
             /* @var AssessmentTestSession $session */
             $session = $context->getTestSession();
             $route = $session->getRoute();
             $section = $session->getCurrentAssessmentSection();
-            $limits = $section->getTimeLimits();
-
-            // As we have only one identifier for the whole adaptive section it will consider a jump of section on the
-            // first item
-            if (!($context instanceof QtiRunnerServiceContext) || !$context->isAdaptive()) {
-                $isJumpOutOfSection = false;
-                if (($nextPosition >= 0) && ($nextPosition < $route->count())) {
-                    $nextSection = $route->getRouteItemAt($nextPosition);
-
-                    // phpcs:disable Generic.Files.LineLength
-                    $isJumpOutOfSection = ($section->getIdentifier() !== $nextSection->getAssessmentSection()->getIdentifier());
-                    // phpcs:enable Generic.Files.LineLength
-                }
 
-                if ($isJumpOutOfSection && $limits != null && $limits->hasMaxTime()) {
-                    $assessmentItemRefs = $section->getComponentsByClassName('assessmentItemRef');
-                    foreach ($assessmentItemRefs as $assessmentItemRef) {
-                        $itemSessions = $session->getAssessmentItemSessions($assessmentItemRef->getIdentifier());
-
-                        if ($itemSessions !== false) {
-                            foreach ($itemSessions as $itemSession) {
-                                $itemSession->endItemSession();
-                            }
-                        }
-                    }
+            // As we have only one identifier for the whole adaptive section
+            // it will consider a jump of section on the first item
+            if (
+                !self::isAdaptive($context)
+                && self::jumpsOutOfSection($route, $section, $nextPosition)
+                && self::isTimeLimited($section)
+            ) {
+                self::endItemSessions($session, $section);
+            }
+        }
+    }
+
+    private static function endItemSessions(
+        AssessmentTestSession $session,
+        AssessmentSection $section
+    ): void {
+        foreach ($section->getComponentsByClassName('assessmentItemRef') as $assessmentItemRef) {
+            $itemSessions = $session->getAssessmentItemSessions($assessmentItemRef->getIdentifier());
+
+            if ($itemSessions !== false) {
+                foreach ($itemSessions as $itemSession) {
+                    $itemSession->endItemSession();
                 }
             }
         }
     }
+
+    private static function jumpsOutOfSection(
```

Comment:
Why private `static` methods? Just out of curiosity, I noticed we are preferring this rather than non-static methods.

Reply:
True. It would be good to have it in a separate class/methods cause we are increasing the static methods there. This is not mandatory, just highlight that refactoring a legacy static method in more static methods might not be optimal.

Reply:
Sure. I'll try to do that, maybe it can be put in some sort of specification class or helper service.

Reply:
No need to to, unless we have time for this. I am not sure we have.